extensions/amp-apester-media/0.1/amp-apester-media.js

```diff
@@ -0,0 +1,141 @@
+/**
+ * Copyright 2015 The AMP HTML Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {CSS} from '../../../build/amp-apester-media-0.1.css';
+import {user} from '../../../src/log';
+import {loadPromise} from '../../../src/event-helper';
+import {setStyles} from '../../../src/style';
+import {listenFor} from '../../../src/iframe-helper';
+
+/**
+ * AMP Apester-media
+ */
+class AmpApesterMedia extends AMP.BaseElement {
+
+  // /** @override */
+  // isLayoutSupported() {
+  //   return true;// (layout == Layout.CONTAINER);
+  // }
+
+  /** @override */
+  viewportCallback() {
+    // TODO: if amp left the viewport, send iframe post message that it has been seen.
+  }
+
+  /** @override */
+  buildCallback() {
+    /**
+     * @private {?String}
+     */
+    this.displayBaseUrl_ = window.location.protocol +
+      '//stage3-renderer.qmerce.com'; //renderer.qmerce.local.com
+
+    /**
+     * @private {?Element}
+     */
+    this.mediaAttribute_ = user.assert(
+      (this.element.getAttribute('data-apester-media-id') ||
+      this.element.getAttribute('data-apester-channel-token')),
+      'Either the data-apester-media-id or the data-apester-channel-token ' +
+      'attributes must be specified for <amp-apester-media> %s',
+      this.element);
+
+    /**
+     * @private {?Element}
+     */
+    this.iframe_ = null;
+    /**
+     * @private {?Promise}
+     */
+    this.iframePromise_ = null;
+
+    //TODO change to setStyle?
+    this.element.classList.add('-amp-apester-container');
+  }
+
+  /** @override */
+  firstLayoutCompleted() {
+    // Do not hide placeholder
+  }
+
+  /** @override */
+  layoutCallback() {
```

Comment:
Please also implement `unlayoutCallback` and possibly `pauseCallback`. See other similar extensions (Instagram, YouTube).

Basically your extension needs to remove heavy resources (e.g. iframes) it created when unlayoutCallback is called. This is probably true for all kind of embeddables in apster - except for videos.

If apster have video embeds, `pauseCallback` should implement the functionality (possibly through `postMessage` calls) to pause videos if they are playing. If apster doesn't expose a pausing functionality then `unlayoutOnPause() => true` to destroy the video iframe should be used.

Since apster has "progression" in using the widgets, this might be up for debates, for example, if a poll has 10 "slides" of questions and the user navigated away on question 5, maybe AMP should still keep the poll and not destroy it to allow the user to complete the quiz if they navigate back. Not sure here, cc/ @dvoytenko